### PR TITLE
Builds: set scale-in protection before/after each build

### PR DESCRIPTION
**Pull request by humitos**
We use AWS scale-in protection to protect our instances from being scaled-in automatically while running a build. This avoids killing the build in the middle while it's running.

Besides, it could help us to be more aggressive to scale-in if we consider without degrading user experience.

Related https://github.com/readthedocs/readthedocs-ops/issues/1324
Requires: https://github.com/readthedocs/readthedocs-ops/pull/1356

### CODE REVIEW

**Review by stsewd — COMMENTED** [review]
LGTM, you may want a review from someone else too, I'm not that familiar to know if there are any side effects from this.

File: readthedocs/projects/tasks/utils.py [hunk 1]
```diff
@@ -321,6 +323,49 @@ def deprecated_config_file_used_notification():
...
+
+    asg = boto3.client(
+        "autoscaling",
+        aws_access_key_id=settings.RTD_AWS_SCALE_IN_ACCESS_KEY,
+        aws_secret_access_key=settings.RTD_AWS_SCALE_IN_SECRET_ACCESS_KEY,
+        region_name=settings.RTD_AWS_SCALE_IN_REGION_NAME,
+    )
+
+    # web-extra-i-0c3e866c4e323928f
+    hostname_match = re.match(r"([a-z\-]+)-(i-[a-z0-9]+)", instance)
```
Comment (agjohnson) [inline readthedocs/projects/tasks/utils.py:350]: Not likely that any match would be greedier than we want here, but EC2 identifiers are hex encoded:

```suggestion
    hostname_match = re.match(r"([a-z\-]+)-(i-[a-f0-9]+)", instance)
```

**Review by agjohnson — COMMENTED** [review]
The change here is looking good, though I might hesitate to turn this on for all projects all at once. If feature flag is too much, we can always test this with a partial deploy or something, but I suspect we might want to leave this on for a while to test this all works as expected.

**Comment by humitos** [comment]
I put the scale-in protection under a feature flag. It's ready for re-review.

File: readthedocs/projects/tasks/builds.py [hunk 2]
```diff
@@ -728,6 +744,13 @@ def after_return(self, status, retval, task_id, args, kwargs, einfo):
         except Exception:
             log.exception("Failed to revoke build api key.", exc_info=True)
 
+        # Disable scale-in protection on this instance
+        if self.data.project.has_feature(Feature.SCALE_IN_PROTECTION):
+            set_builder_scale_in_protection.delay(
+                instace=socket.gethostname(),
```
Comment (agjohnson) [inline readthedocs/projects/tasks/builds.py:750]: Oops, sorry I wasn't clear, there is a typo here too though.

**Review by agjohnson — APPROVED** [review]
This looks great. I'm excited to see what exactly this does for us :crossed_fingers:

File: readthedocs/projects/tasks/builds.py (hunk 2 above)
Comment (humitos) [reply readthedocs/projects/tasks/builds.py:750]: Meh, you where clear but I think I made a mistake 🙃

Comment (humitos) [reply readthedocs/projects/tasks/builds.py:750]: Fixed

**Pull request merged**